File: experiments/toy-agent/python/agent.py
"""A toy coding agent in ~80 lines, using the official Anthropic SDK.

The whole agent is `run_turn`: call the model, run any tools it asks for, feed
the results back, repeat until it stops asking. That loop IS the agent.
"""

import os
import anthropic

MODEL = "claude-opus-4-8"
client = anthropic.Anthropic()  # reads ANTHROPIC_API_KEY from the environment

# --- Tool definitions: the only things the agent can do in the world ----------
TOOLS = [
    {
        "name": "list_files",
        "description": "List the files and directories in a directory. "
                       "Call this to explore the codebase before reading files.",
        "input_schema": {
            "type": "object",
            "properties": {
                "directory": {
                    "type": "string",
                    "description": "Directory to list. Defaults to the current directory.",
                },
            },
        },
    },
    {
        "name": "read_file",
        "description": "Read the full contents of a file. Call this when you "
                       "need to see what is inside a specific file.",
        "input_schema": {
            "type": "object",
            "properties": {
                "path": {"type": "string", "description": "Path to the file to read."},
            },
            "required": ["path"],
        },
    },
]


# --- Tool implementations: your code decides what actually happens -------------
def list_files(directory="."):
    try:
        return "\n".join(sorted(os.listdir(directory)))
    except OSError as e:
        return f"error: {e}"


def read_file(path):
    try:
        with open(path, "r", encoding="utf-8") as f:
            return f.read()
    except OSError as e:
        return f"error: {e}"


def execute_tool(name, tool_input):
    if name == "list_files":
        return list_files(tool_input.get("directory", "."))
    if name == "read_file":
        return read_file(tool_input.get("path", ""))
    return f"error: unknown tool {name}"


# --- The loop ------------------------------------------------------------------
def run_turn(messages):
    while True:
        response = client.messages.create(
            model=MODEL,
            max_tokens=16000,
            tools=TOOLS,
            messages=messages,
        )
        # Echo the assistant's turn (text + tool_use blocks) back into history.
        messages.append({"role": "assistant", "content": response.content})

        for block in response.content:
            if block.type == "text":
                print(block.text)

        tool_uses = [b for b in response.content if b.type == "tool_use"]
        if not tool_uses:
            return  # no tool requested -> the turn is done

        tool_results = [
            {
                "type": "tool_result",
                "tool_use_id": tu.id,
                "content": execute_tool(tu.name, tu.input),
            }
            for tu in tool_uses
        ]
        messages.append({"role": "user", "content": tool_results})


def main():
    messages = []
    print("Toy agent. Ask about files in this directory. Ctrl-D to quit.")
    while True:
        try:
            user = input("you> ")
        except EOFError:
            break
        messages.append({"role": "user", "content": user})
        run_turn(messages)


if __name__ == "__main__":
    main()

File: experiments/toy-agent/clojure/src/agent.clj
;; A toy coding agent in Clojure, talking to the raw HTTP API (no SDK).
;;
;; The loop is `run`, a tail-recursive function over the growing message vector:
;; call the model, run any tools it asks for, conj the results on, and `recur`
;; until it stops asking. Because clj-http parses the JSON into Clojure data and
;; cheshire serializes it back, echoing the assistant turn is just (:content response).

(ns agent
  (:require [clj-http.client :as http]
            [cheshire.core :as json]
            [clojure.java.io :as io]
            [clojure.string :as string]))

(def model "claude-opus-4-8")
(def url "https://api.anthropic.com/v1/messages")
(def api-key (System/getenv "ANTHROPIC_API_KEY"))

;; --- Tool definitions: the only things the agent can do in the world ---------
(def tools
  [{:name "list_files"
    :description (str "List the files and directories in a directory. "
                      "Call this to explore the codebase before reading files.")
    :input_schema {:type "object"
                   :properties {:directory {:type "string"
                                            :description "Directory to list. Defaults to the current directory."}}}}
   {:name "read_file"
    :description (str "Read the full contents of a file. Call this when you "
                      "need to see what is inside a specific file.")
    :input_schema {:type "object"
                   :properties {:path {:type "string" :description "Path to the file to read."}}
                   :required ["path"]}}])

;; --- Tool implementations ----------------------------------------------------
(defn execute-tool [name input]
  (try
    (case name
      "list_files" (->> (.listFiles (io/file (get input :directory ".")))
                        (map #(.getName %))
                        sort
                        (string/join "\n"))
      "read_file"  (slurp (:path input))
      (str "error: unknown tool " name))
    (catch Exception e (str "error: " (.getMessage e)))))

;; --- One round trip to the API -----------------------------------------------
(defn call-api [messages]
  (:body (http/post url
                    {:headers {"x-api-key" api-key
                               "anthropic-version" "2023-06-01"}
                     :content-type :json
                     :as :json
                     :body (json/generate-string {:model model
                                                  :max_tokens 16000
                                                  :tools tools
                                                  :messages messages})})))

;; --- The loop ----------------------------------------------------------------
(defn run [messages]
  (let [response  (call-api messages)
        content   (:content response)
        ;; Echo the assistant turn back into history verbatim.
        messages  (conj messages {:role "assistant" :content content})
        tool-uses (filter #(= "tool_use" (:type %)) content)]
    (if (empty? tool-uses)
      (do (doseq [b content]
            (when (= "text" (:type b)) (println (:text b))))
          messages)
      (let [results  (mapv (fn [tu]
                             {:type "tool_result"
                              :tool_use_id (:id tu)
                              :content (execute-tool (:name tu) (:input tu))})
                           tool-uses)
            messages (conj messages {:role "user" :content results})]
        (recur messages)))))

(defn -main [& _]
  (println "Toy agent. Ask about files in this directory. Ctrl-D to quit.")
  (loop [messages []]
    (print "you> ") (flush)
    (when-let [line (read-line)]
      (recur (run (conj messages {:role "user" :content line}))))))
